### PR TITLE
feat(sdk-coin-ada): staking transaction builder

### PR DESCRIPTION
**Pull request by margueriteblair**
Ticket: BG-50797

### CODE REVIEW

File: modules/sdk-coin-ada/src/lib/utils.ts [hunk 1]
```diff
@@ -1,8 +1,57 @@
...
+    stakeKeyPair: KeyPair,
+    paymentKeyPair: KeyPair,
+    network: AddressFormat
+  ): string {
+    let baseAddr;
+    if (network === AddressFormat.mainnet) {
+      // 1. create stake pubKey
+      const key = stakeKeyPair.getKeys().pub;
+
+      const stakePub = Bip32PublicKey.from_bytes(Buffer.from(key));
```
Comment (ewangbitgo) [inline modules/sdk-coin-ada/src/lib/utils.ts:23]: why use bip32 key?

Comment (margueriteblair) [reply modules/sdk-coin-ada/src/lib/utils.ts:23]: When I do the same process, PublicKey.from_bytes(key) it throws an invalid size error

Comment (ewangbitgo) [reply modules/sdk-coin-ada/src/lib/utils.ts]: try Buffer.from(key, 'hex') as key is in hex format

File: modules/sdk-coin-ada/test/unit/transactionBuilder.ts [hunk 2]
```diff
@@ -203,4 +205,270 @@ describe('ADA Transaction Builder', async () => {
...
+    const txBuilder = factory.getStakingWithdrawBuilder();
+    const utxoData = await axios.get('https://testnet.koios.rest/api/v0/address_info?_address=' + senderAddress);
+    const senderBalance = utxoData.data[0].balance;
+    const utxoSet = utxoData.data[0].utxo_set;
+    for (const utxo of utxoSet) {
+      txBuilder.input({ transaction_id: utxo.tx_hash, transaction_index: utxo.tx_index });
+    }
+    txBuilder.changeAddress(senderAddress, senderBalance);
+
+    const rewardAcctData = await axios.get('https://testnet.koios.rest/api/v0/account_info?_address=' + senderAddress);
```
Comment (ewangbitgo) [inline modules/sdk-coin-ada/test/unit/transactionBuilder.ts:410]: change it to staking address derived from staking key

File: modules/sdk-coin-ada/test/unit/transactionBuilder.ts [hunk 3]
```diff
@@ -203,4 +205,270 @@ describe('ADA Transaction Builder', async () => {
...
+  });
+
+  it('should withdraw rewards from a stake address', async () => {
+    const axiosConfig = {
+      headers: {
+        'Content-Type': 'application/cbor',
+      },
+      timeout: 10000,
+    };
+    const senderAddress = testData.stakingAccount.baseAddress;
```
Comment (ewangbitgo) [inline modules/sdk-coin-ada/test/unit/transactionBuilder.ts:398]: senderaddress has to derived from payment key

File: modules/sdk-coin-ada/test/unit/transactionBuilder.ts [hunk 4]
```diff
@@ -203,4 +205,270 @@ describe('ADA Transaction Builder', async () => {
...
+    const utxoSet = utxoData.data[0].utxo_set;
+    for (const utxo of utxoSet) {
+      txBuilder.input({ transaction_id: utxo.tx_hash, transaction_index: utxo.tx_index });
+    }
+    txBuilder.changeAddress(senderAddress, senderBalance);
+
+    const rewardAcctData = await axios.get('https://testnet.koios.rest/api/v0/account_info?_address=' + senderAddress);
+    const stakeAddressBalance = rewardAcctData.data[0].rewards;
+    console.log(stakeAddressBalance);
+    txBuilder.addWithdrawal(senderAddress, stakeAddressBalance, AddressFormat.testnet);
```
Comment (ewangbitgo) [inline modules/sdk-coin-ada/test/unit/transactionBuilder.ts:413]: do withdraw need full address or just reward address?

File: modules/sdk-coin-ada/src/lib/stakingWithdrawRewardsBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  protected fromImplementation(rawTransaction: any): Transaction {
+    return super.fromImplementation(rawTransaction);
+  }
+
+  /** @inheritdoc */
+  protected signImplementation(key: BaseKey): Transaction {
+    return super.signImplementation(key);
+  }
+
+  addWithdrawal(stakingPubKey: string, value: string) {
```
Comment (margueriteblair) [inline modules/sdk-coin-ada/src/lib/stakingWithdrawRewardsBuilder.ts:32]: why remove network check?

File: modules/sdk-coin-ada/test/unit/StakingWithdrawRewardsBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  it('should submit a staking withdraw using signature interface', async () => {
+    const keyPairPayment = new KeyPair({ prv: testData.privateKeys.prvKey9 });
+    const keyPairStake = new KeyPair({ prv: testData.privateKeys.prvKey10 });
+    const senderAddress = Utils.default.createBaseAddressWithStakeAndPaymentKey(
+      keyPairStake,
+      keyPairPayment,
+      AddressFormat.testnet
+    );
+    const txBuilder = factory.getStakingWithdrawBuilder();
+    const senderBalance = '22122071';
```
Comment (margueriteblair) [inline modules/sdk-coin-ada/test/unit/StakingWithdrawRewardsBuilder.ts:114]: Why hardcoded?

Comment (ewangbitgo) [reply modules/sdk-coin-ada/test/unit/StakingWithdrawRewardsBuilder.ts:114]: this test is for offline test

File: modules/sdk-coin-ada/src/lib/stakingWithdrawRewardsBuilder.ts (hunk 5 above)
Comment (ewangbitgo) [reply modules/sdk-coin-ada/src/lib/stakingWithdrawRewardsBuilder.ts:32]: i changed input parameter to staking key and value

File: modules/sdk-coin-ada/src/lib/stakingWithdrawRewardsBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    const stakePub = CardanoWasm.PublicKey.from_bytes(Buffer.from(stakingPubKey, 'hex'));
+    const rewardAddress = CardanoWasm.RewardAddress.new(
+      CardanoWasm.NetworkInfo.testnet().network_id(),
```
Comment (ewangbitgo) [inline modules/sdk-coin-ada/src/lib/stakingWithdrawRewardsBuilder.ts:35]: needs to check if it is mainnet

File: modules/sdk-coin-ada/src/lib/transaction.ts [hunk 8]
```diff
@@ -169,11 +238,24 @@ export class Transaction extends BaseTransaction {
...
+    changeAmount: string;
+    type: string;
+    withdrawals: Withdrawal[];
+  } {
     const txJson = this.toJson();
     const displayOrder = ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee', 'type'];
     const amount = txJson.outputs.map((o) => ({ amount: BigInt(o.amount) }));
     const outputAmount = amount.reduce((p, n) => p + BigInt(n.amount), BigInt('0')).toString();
+    const type =
+      txJson.certs.length === 0 ? (txJson.withdrawals.length === 0 ? 'Transfer' : 'StakeWithdarw') : 'StakingActivate';
```
Comment (margueriteblair) [inline modules/sdk-coin-ada/src/lib/transaction.ts:258]: typo in StakeWithdraw

File: modules/sdk-coin-ada/src/lib/transactionBuilder.ts [hunk 9]
```diff
@@ -183,15 +210,47 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
         )
       );
     });
-
     if (this._changeAddress && this._senderBalance) {
       const changeAddress = CardanoWasm.Address.from_bech32(this._changeAddress);
       const utxoBalance = CardanoWasm.BigNum.from_str(this._senderBalance);
-      const change = utxoBalance.checked_sub(fee).checked_sub(totalAmountToSend);
+
+      const adjustment = 2000000;
```
Comment (margueriteblair) [inline modules/sdk-coin-ada/src/lib/transactionBuilder.ts:217]: nit: we can change adjustment to already be BigNum.from_str('2000000') to reduce redundancy

File: modules/sdk-coin-ada/src/lib/transaction.ts [hunk 10]
```diff
@@ -169,11 +238,24 @@ export class Transaction extends BaseTransaction {
...
+    changeAmount: string;
+    type: string;
+    withdrawals: Withdrawal[];
+  } {
     const txJson = this.toJson();
     const displayOrder = ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee', 'type'];
     const amount = txJson.outputs.map((o) => ({ amount: BigInt(o.amount) }));
     const outputAmount = amount.reduce((p, n) => p + BigInt(n.amount), BigInt('0')).toString();
+    const type =
+      txJson.certs.length === 0 ? (txJson.withdrawals.length === 0 ? 'Transfer' : 'StakeWithdraw') : 'StakingActivate';
```
Comment (ewangbitgo) [inline modules/sdk-coin-ada/src/lib/transaction.ts:258]: missing staking deactivate, maybe we should get it form this._type directly

**Review by ewangbitgo — DISMISSED** [review]

File: modules/sdk-coin-ada/src/lib/transaction.ts [hunk 11]
```diff
@@ -152,6 +206,21 @@ export class Transaction extends BaseTransaction {
       this._transaction = txn;
       this._id = Buffer.from(CardanoWasm.hash_transaction(txn.body()).to_bytes()).toString('hex');
       this._type = TransactionType.Send;
+      if (this._transaction.body().certs()) {
+        for (let i = 0; i < this._transaction.body().certs()!.len(); i++) {
+          const cert = this._transaction.body().certs()!.get(i);
+          if (cert.as_stake_registration() !== undefined) {
+            this._type = TransactionType.StakingActivate;
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/transaction.ts:213]: Is there value in `break`ing the loop once `this_type` is assigned?

File: modules/sdk-coin-ada/src/lib/transaction.ts [hunk 12]
```diff
@@ -99,6 +113,47 @@ export class Transaction extends BaseTransaction {
       });
     }
 
+    if (this._transaction.body().certs()) {
+      for (let i = 0; i < this._transaction.body().certs()!.len(); i++) {
+        const cert = this._transaction.body().certs()!.get(i);
+        if (cert.as_stake_registration() !== undefined) {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/transaction.ts:119]: Same question here about breaking the loop on assignment.

File: modules/sdk-coin-ada/src/lib/transaction.ts [hunk 13]
```diff
@@ -152,6 +206,21 @@ export class Transaction extends BaseTransaction {
...
+          const cert = this._transaction.body().certs()!.get(i);
+          if (cert.as_stake_registration() !== undefined) {
+            this._type = TransactionType.StakingActivate;
+          }
+          if (cert.as_stake_deregistration() !== undefined) {
+            this._type = TransactionType.StakingDeactivate;
+          }
+        }
+      }
+      if (this._transaction.body().withdrawals()) {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/transaction.ts:220]: Should this be an `else` or is it possible that the `certs` loop will be overwritten by this `if` statement?

File: modules/sdk-coin-ada/src/lib/transaction.ts [hunk 14]
```diff
@@ -169,11 +238,30 @@ export class Transaction extends BaseTransaction {
   }
 
   /** @inheritdoc */
-  explainTransaction(): TransactionExplanation {
+  explainTransaction(): {
+    outputs: { amount: string; address: string }[];
+    certificates: Cert[];
+    changeOutputs: any[];
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/transaction.ts:244]: Can we get `any` defined here?

File: modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  protected get transactionType(): TransactionType {
+    return TransactionType.StakingActivate;
+  }
+
+  /** @inheritdoc */
+  initBuilder(tx: Transaction): void {
+    super.initBuilder(tx);
+  }
+
+  stakingCredential(stakingPublicKey: string, poolHash: string): this {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts:22]: Can we get the jsdocs for this function?

File: modules/sdk-coin-ada/src/lib/stakingDeactivateBuilder.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+  constructor(_coinConfig: Readonly<CoinConfig>) {
+    super(_coinConfig);
+    this._type = TransactionType.StakingDeactivate;
+  }
+
+  protected get transactionType(): TransactionType {
+    return TransactionType.StakingDeactivate;
+  }
+
+  stakingCredential(stakingPublicKey: string): this {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/stakingDeactivateBuilder.ts:18]: Can we get the jsdocs for this function?

File: modules/sdk-coin-ada/src/lib/stakingWithdrawRewardsBuilder.ts [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  protected fromImplementation(rawTransaction: any): Transaction {
+    return super.fromImplementation(rawTransaction);
+  }
+
+  /** @inheritdoc */
+  protected signImplementation(key: BaseKey): Transaction {
+    return super.signImplementation(key);
+  }
+
+  addWithdrawal(stakingPubKey: string, value: string) {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/stakingWithdrawRewardsBuilder.ts:30]: Can we get the jsdocs for this function?

**Review by mmcshinsky-bitgo — CHANGES_REQUESTED** [review]

File: modules/sdk-coin-ada/src/lib/transaction.ts (hunk 11 above)
Comment (margueriteblair) [reply modules/sdk-coin-ada/src/lib/transaction.ts:213]: meh, not really imo, certs array will only be like 2 certs long max

File: modules/sdk-coin-ada/src/lib/transaction.ts (hunk 12 above)
Comment (margueriteblair) [reply modules/sdk-coin-ada/src/lib/transaction.ts:119]: yeah, same as above, not huge reason to break

File: modules/sdk-coin-ada/src/lib/transaction.ts (hunk 13 above)
Comment (margueriteblair) [reply modules/sdk-coin-ada/src/lib/transaction.ts:220]: This certs loop should be overwritten here if this is the case!

File: modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts (hunk 15 above)
Comment (margueriteblair) [reply modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts:22]: Oh man...forgive me, what does jsdocs refer to here?

Comment (mmcshinsky-bitgo) [reply modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts:22]: Not a problem at all. Here is an example: https://github.com/BitGo/BitGoJS/blob/master/modules/sdk-coin-ada/src/ada.ts#L77
Here is the "official" source: https://jsdoc.app/index.html

Comment (mmcshinsky-bitgo) [reply modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts:22]: That one is missing the types. Should be like `@param {string} variableName <Description>`

**Review by alia-bitgo — DISMISSED** [review]

File: modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts (hunk 15 above)
Comment (mmcshinsky-bitgo) [reply modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts:28]: nit: missing curly braces. Not a blocker but should be fixed in this or another PR.

File: modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+  /** @inheritdoc */
+  protected async buildImplementation(): Promise<Transaction> {
+    const tx = await super.buildImplementation();
+    tx.setTransactionType(TransactionType.StakingActivate);
+    return tx;
+  }
+
+  /** @inheritdoc */
+  protected fromImplementation(rawTransaction: any): Transaction {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/stakingActivateBuilder.ts:54]: Can this `any` be defined?

File: modules/sdk-coin-ada/src/lib/stakingDeactivateBuilder.ts [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+  /** @inheritdoc */
+  protected async buildImplementation(): Promise<Transaction> {
+    const tx = await super.buildImplementation();
+    tx.setTransactionType(TransactionType.StakingDeactivate);
+    return tx;
+  }
+
+  /** @inheritdoc */
+  protected fromImplementation(rawTransaction: any): Transaction {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-ada/src/lib/stakingDeactivateBuilder.ts:42]: Same here.

**Review by alia-bitgo — APPROVED** [review]

**Review by mmcshinsky-bitgo — APPROVED** [review]

**Pull request merged**